File: app/src/main/java/com/qweex/smashuphelper/Persistable.java
package com.qweex.smashuphelper;


import android.content.Context;
import android.content.SharedPreferences;
import android.os.Parcel;
import android.os.Parcelable;
import android.util.Log;

import java.io.Serializable;

public class Persistable implements Serializable, Parcelable {
    public String name;
    private boolean disabled;
    final static private String PREF_NAME = "persistent";
    private static SharedPreferences prefs;

    Persistable(String key) {
        name = key;
        disabled = prefs.getBoolean(String.format("%s[%s]", getClass().getName(), name), true);
        Log.d("Persistable", key + " " + disabled);
    }

    Persistable(Parcel src) {
        this(src.readString());
        setDisabled(src.readByte() != 0);
    }

    public void setDisabled(boolean d) {
        disabled = d;
        prefs.edit().putBoolean(
                String.format("%s[%s]", getClass().getName(), name),
                disabled
        ).apply();
    }

    public boolean isDisabled() {
        return disabled;
    }

    @Override
    public int describeContents() {
        return 0;
    }

    @Override
    public void writeToParcel(Parcel dest, int flags) {
        dest.writeString(name);
        dest.writeByte((byte) (isDisabled() ? 1 : 0));
    }


    public static Parcelable.Creator<Persistable> CREATOR = new Parcelable.Creator<Persistable>() {
        @Override
        public Persistable createFromParcel(Parcel in) {
            return new Persistable(in);
        }

        @Override
        public Persistable[] newArray(int size) {
            return new Persistable[size];
        }
    };

    public static void init(Context c) {
        prefs = c.getSharedPreferences(PREF_NAME, Context.MODE_PRIVATE);
    }
}

File: smashuphelper/src/main/java/com/qweex/smashuphelper/Persistable.kt
package com.qweex.smashuphelper


import android.annotation.SuppressLint
import android.content.Context
import android.content.Context.MODE_PRIVATE
import android.content.SharedPreferences
import android.os.Parcel
import android.os.Parcelable
import android.util.Log
import java.io.Serializable


@SuppressLint("ParcelCreator")
open class Persistable internal constructor(var name: String) : Serializable, Parcelable {
    private var disabled: Boolean = false

    private fun prefKey(): String {
        return String.format("%s[%s]", javaClass.name, name)
    }

    var isDisabled: Boolean
        get() = disabled
        set(d) {
            disabled = d
            Log.d("Persisting", prefKey() + " = " + d)
            prefs!!.edit().putBoolean(prefKey(), disabled).apply()
        }

    init {
        disabled = prefs!!.getBoolean(prefKey(), false)
        Log.d("Persistable", "$name $disabled")
    }

    internal constructor(src: Parcel) : this(src.readString()) {
        val x = src.readByte()
        Log.d("Persistparcle", "$name=$x")
        isDisabled = x.toInt() != 0
    }

    override fun describeContents(): Int {
        return 0
    }

    override fun writeToParcel(dest: Parcel, flags: Int) {
        Log.d("PersistparcleW", "$name=$disabled")
        dest.writeString(name)
        dest.writeByte((if (disabled) 1 else 0).toByte())
    }

    companion object CREATOR : Parcelable.Creator<Persistable> {
        override fun createFromParcel(parcel: Parcel): Persistable {
            return Persistable(parcel)
        }

        override fun newArray(size: Int): Array<Persistable?> {
            return arrayOfNulls(size)
        }
        private var prefs: SharedPreferences? = null
        private const val PREF_NAME = "persistent"
        fun init(c: Context)
        {
            prefs = c.getSharedPreferences(PREF_NAME, MODE_PRIVATE)
        }
    }

}
